core/chains/evm/forwarders/forwarder_manager.go

```diff
@@ -0,0 +1,308 @@
+			continue
+		}
+		for _, EOA := range EOAs {
+			if EOA == from {
+				forwardedPayload, err := f.getForwardedPayload(to, encodedPayload)
+				if err != nil {
+					f.logger.AssumptionViolation("Forwarder encoding failed, this should never happen")
+					continue
+				}
+				return fwdr.Address, forwardedPayload, nil
+			}
+		}
+	}
+
+	return to, encodedPayload, errors.New("Skipping forwarding transaction")
+}
+
+func (f *FwdMgr) getForwardedPayload(dest common.Address, origPayload []byte) ([]byte, error) {
+	callArgs, err := forwardABI.Inputs.Pack(dest, origPayload)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to pack forwarder payload")
+	}
+
+	dataBytes := append(forwardABI.ID, callArgs...)
+	return dataBytes, nil
+}
+
+func (f *FwdMgr) getContractSenders(addr common.Address) ([]common.Address, error) {
+	if senders, ok := f.getCachedSenders(addr); ok {
+		return senders, nil
+	}
+	senders, err := f.getAuthorizedSenders(addr)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to call getAuthorizedSenders on %s", addr)
+	}
+	f.setCachedSenders(addr, senders)
+	f.subscribeSendersChangedLogs(addr)
+	return senders, nil
+}
+
+func (f *FwdMgr) getAuthorizedSenders(addr common.Address) ([]common.Address, error) {
+	c, err := authorized_receiver.NewAuthorizedReceiverCaller(addr, f.evmClient)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to init forwarder caller")
+	}
+	opts := bind.CallOpts{Context: f.ctx, Pending: false}
+	senders, err := c.GetAuthorizedSenders(&opts)
+	if err != nil {
+		return nil, err
+	}
+	return senders, nil
+}
+
+func (f *FwdMgr) initForwardersCache(ctx context.Context, fwdrs []Forwarder) {
+	for _, fwdr := range fwdrs {
+		senders, err := f.getAuthorizedSenders(fwdr.Address)
+		if err != nil {
+			f.logger.Warnw("Failed to call getAuthorizedSenders on forwarder", fwdr, "err", err)
+			continue
+		}
+		f.setCachedSenders(fwdr.Address, senders)
+
+	}
+}
+func (f *FwdMgr) subscribeForwardersLogs(fwdrs []Forwarder) {
+	for _, fwdr := range fwdrs {
+		f.subscribeSendersChangedLogs(fwdr.Address)
+	}
+}
+
+func (f *FwdMgr) subscribeSendersChangedLogs(addr common.Address) {
+	f.logpoller.MergeFilter(
+		[]common.Hash{authorized_receiver.AuthorizedReceiverAuthorizedSendersChanged{}.Topic()},
+		addr)
+}
+
+func (f *FwdMgr) setCachedSenders(addr common.Address, senders []common.Address) {
+	f.cacheMu.Lock()
+	defer f.cacheMu.Unlock()
+	f.sendersCache[addr] = senders
+}
+
+func (f *FwdMgr) getCachedSenders(addr common.Address) ([]common.Address, bool) {
+	f.cacheMu.RLock()
+	defer f.cacheMu.RUnlock()
+	addrs, ok := f.sendersCache[addr]
+	return addrs, ok
+}
+
+func (f *FwdMgr) runLoop() {
+	defer f.wg.Done()
+	tick := time.After(0)
+
+	for {
+		select {
+		case <-tick:
+			tick = time.After(utils.WithJitter(authLogCheckInterval))
+			addrs := f.collectAddresses()
+			logs, err := f.logpoller.LatestLogEventSigsAddrsWithConfs(f.latestBlock, authTopics, addrs, 0)
```

Comment:
hmm wait if we're polling for changed logs periodically then extracting the new senders, couldn't we just poll the contract state itself (just a state read so it'd be fast)?

Reply:
this call should do 0 work ~99% of the time. 5 second is a super cautious ticker value and I actually want to increase that by quite a bit, maybe once every 5-10 mins call is enough.